### PR TITLE
Docs/add decomposition complexities to several multiplexor based boxes

### PR DESCRIPTION
**Pull request by yao-cqc**
# Description

Add gate complexities to all the `Multiplexed*Boxes `, `StatePreparationBox`, and `DiagonalBox`

# Related issues

Please mention any github issues addressed by this PR.

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: pytket/binders/circuit/boxes.cpp [hunk 1]
```diff
@@ -981,7 +981,11 @@ void init_boxes(py::module &m) {
       "A user-defined multiplexed rotation gate (i.e. "
       "uniformly controlled single-axis rotations) specified by "
       "a map from bitstrings to " CLSOBJS(Op)
-      "or a list of bitstring-" CLSOBJS(Op) " pairs")
+      "or a list of bitstring-" CLSOBJS(Op) " pairs. "
+      "Implamentation based on https://arxiv.org/abs/quant-ph/0410066. "
```
Comment (sjdilkes) [inline pytket/binders/circuit/boxes.cpp:985]: "Implementation based on arxiv.org/abs/quant-ph/0410066. "

File: pytket/binders/circuit/boxes.cpp [hunk 2]
```diff
@@ -1048,7 +1052,11 @@ void init_boxes(py::module &m) {
       "A user-defined multiplexed U2 gate (i.e. uniformly controlled U2 "
       "gate) specified by a "
       "map from bitstrings to " CLSOBJS(Op)
-      "or a list of bitstring-" CLSOBJS(Op) " pairs")
+      "or a list of bitstring-" CLSOBJS(Op) " pairs"
+      "Implamentation based on https://arxiv.org/abs/quant-ph/0410066. "
```
Comment (sjdilkes) [inline pytket/binders/circuit/boxes.cpp:1056]: "Implementation based on arxiv.org/abs/quant-ph/0410066. "

File: pytket/binders/circuit/boxes.cpp [hunk 3]
```diff
@@ -1132,7 +1147,15 @@ void init_boxes(py::module &m) {
   py::class_<StatePreparationBox, std::shared_ptr<StatePreparationBox>, Op>(
       m, "StatePreparationBox",
       "A box for preparing quantum states using multiplexed-Ry and "
-      "multiplexed-Rz gates")
+      "multiplexed-Rz gates. "
+      "Implementation based on Theorem 9 of "
+      "https://arxiv.org/abs/quant-ph/0406176. "
```
Comment (sjdilkes) [inline pytket/binders/circuit/boxes.cpp:1152]: arxiv.org/abs/quant-ph/0406176.

File: pytket/binders/circuit/boxes.cpp [hunk 4]
```diff
@@ -1162,7 +1185,10 @@ void init_boxes(py::module &m) {
   py::class_<DiagonalBox, std::shared_ptr<DiagonalBox>, Op>(
       m, "DiagonalBox",
       "A box for synthesising a diagonal unitary matrix into a sequence of "
-      "multiplexed-Rz gates.")
+      "multiplexed-Rz gates. "
+      "Implementation based on Theorem 7 of "
+      "https://arxiv.org/abs/quant-ph/0406176. "
```
Comment (sjdilkes) [inline pytket/binders/circuit/boxes.cpp:1190]: arxiv.org/abs/quant-ph/0406176

File: pytket/pytket/_tket/circuit.pyi [hunk 5]
```diff
@@ -2705,7 +2705,7 @@ class CustomGateDef:
         """
 class DiagonalBox(Op):
     """
-    A box for synthesising a diagonal unitary matrix into a sequence of multiplexed-Rz gates.
+    A box for synthesising a diagonal unitary matrix into a sequence of multiplexed-Rz gates. Implementation based on Theorem 7 of https://arxiv.org/abs/quant-ph/0406176. The decomposed circuit has at most 2^n-2 CX gates.
```
Comment (sjdilkes) [inline pytket/pytket/_tket/circuit.pyi:2708]: A box for synthesising an n-qubit diagonal unitary matrix into a sequence of multiplexed-Rz gates. Implementation based on Theorem 7 of arxiv.org/abs/quant-ph/0406176. The decomposed circuit has at most 2^n-2 CX gates.

File: pytket/pytket/_tket/circuit.pyi [hunk 6]
```diff
@@ -2839,7 +2839,7 @@ class MultiBitOp(ClassicalEvalOp):
         """
 class MultiplexedRotationBox(Op):
     """
-    A user-defined multiplexed rotation gate (i.e. uniformly controlled single-axis rotations) specified by a map from bitstrings to :py:class:`Op` sor a list of bitstring-:py:class:`Op` s pairs
+    A user-defined multiplexed rotation gate (i.e. uniformly controlled single-axis rotations) specified by a map from bitstrings to :py:class:`Op` sor a list of bitstring-:py:class:`Op` s pairs. Implamentation based on https://arxiv.org/abs/quant-ph/0410066. The decomposed circuit has at most 2^k single-qubit rotations, 2^k CX gates, and 2 H gates if it's X-axis. k is the number of control qubits.
```
Comment (sjdilkes) [inline pytket/pytket/_tket/circuit.pyi:2842]: A user-defined multiplexed rotation gate (i.e. uniformly controlled single-axis rotations) specified by a map from bitstrings to :py:class:`Op` sor a list of bitstring-:py:class:`Op` s pairs. Implementation based on arxiv.org/abs/quant-ph/0410066. The decomposed circuit has at most 2^k single-qubit rotations, 2^k CX gates, and two additional H gates if the rotation basis is X. k is the number of control qubits.

File: pytket/pytket/_tket/circuit.pyi [hunk 7]
```diff
@@ -2911,7 +2911,7 @@ class MultiplexedTensoredU2Box(Op):
         """
 class MultiplexedU2Box(Op):
     """
-    A user-defined multiplexed U2 gate (i.e. uniformly controlled U2 gate) specified by a map from bitstrings to :py:class:`Op` sor a list of bitstring-:py:class:`Op` s pairs
+    A user-defined multiplexed U2 gate (i.e. uniformly controlled U2 gate) specified by a map from bitstrings to :py:class:`Op` sor a list of bitstring-:py:class:`Op` s pairsImplamentation based on https://arxiv.org/abs/quant-ph/0410066. The decomposed circuit has at most 2^k single-qubit gates, 2^k -1 CX gates, and a k+1 qubit DiagonalBox at the end. k is the number of control qubits.
```
Comment (sjdilkes) [inline pytket/pytket/_tket/circuit.pyi:2914]: pairs. The implementation is based on arxiv.org/abs/quant-ph/0410066. The decomposed circuit has at most 2^k single-qubit gates, 2^k -1 CX gates, and a k+1 qubit DiagonalBox at the end. k is the number of control qubits.

File: pytket/pytket/_tket/circuit.pyi [hunk 8]
```diff
@@ -3686,7 +3686,7 @@ class StabiliserAssertionBox(Op):
         """
 class StatePreparationBox(Op):
     """
-    A box for preparing quantum states using multiplexed-Ry and multiplexed-Rz gates
+    A box for preparing quantum states using multiplexed-Ry and multiplexed-Rz gates. Implementation based on Theorem 9 of https://arxiv.org/abs/quant-ph/0406176. The decomposed circuit has at most 2*(2^n-2) CX gates, and 2^n-2 CX gates if the coefficients are all real. Note that by decomposing each multiplexed-Ry(Rz) as its dagger can lead to a CX cancellation with its adjacent multiplexed-Rz(Ry), which leads to an improved CX count of 2^(n+1)-2n. However, this is currently not automatically done in tket.
```
Comment (sjdilkes) [inline pytket/pytket/_tket/circuit.pyi:3689]: Implementation based on Theorem 9 of arxiv.org/abs/quant-ph/0406176. The decomposed circuit has at most 2*(2^n-2) CX gates, and 2^n-2 CX gates if the coefficients are all real. Note that by decomposing each multiplexed-Ry(Rz) as its dagger can lead to a CX cancellation with its adjacent multiplexed-Rz(Ry), which leads to an improved CX count of 2^(n+1)-2n. However, this is currently not automatically done in tket.

Comment (sjdilkes) [reply pytket/pytket/_tket/circuit.pyi:3689]: I might suggest we remove the note on the improvement though, and instead add a github issue to implement the change?

**Review by sjdilkes — CHANGES_REQUESTED** [review]
Just a few minor edits!

File: pytket/pytket/_tket/circuit.pyi (hunk 8 above)
Comment (yao-cqc) [reply pytket/pytket/_tket/circuit.pyi]: Removed. Github issue is the right place for that note.

**Review by yao-cqc — COMMENTED** [review]
I had some notes to help you find the references, but I did't hit submit... Anyway, I've made the changes you have suggested.

**Review by sjdilkes — APPROVED** [review]
LGTM (& I'm very grateful for the source of truth!)

**Pull request merged**